src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+        let (keys, signatures_threshold) = match credential.value {
+            AccountCredentialWithoutProofs::Initial { icdv } => {
+                (icdv.cred_account.keys, icdv.cred_account.threshold)
+            }
+            AccountCredentialWithoutProofs::Normal { cdv, .. } => {
+                (cdv.cred_key_info.keys, cdv.cred_key_info.threshold)
+            }
+        };
+
+        let mut valid_signatures_count = 0u8;
+
+        for (key_index, public_key) in keys {
+            // If a signature exists for the given credential and key index, verify it and
+            // increase the `valid_signatures_count`.
+            if let Some(cred_sigs) = signatures.sigs.get(&credential_index.index) {
+                if let Some(signature) = cred_sigs.sigs.get(&key_index.0) {
+                    if public_key.verify(message_hash, signature) {
+                        // If the signature is valid, increase the `valid_signatures_count`.
+                        valid_signatures_count += 1;
+                    } else {
+                        // If any signature is invalid, return `false`.
+                        return Ok(false);
+                    }
+                }
+            }
+        }
+
+        // Check if the number of valid signatures meets the required threshold
+        // so that this credential counts as having a valid credential signature.
+        if valid_signatures_count >= signatures_threshold.into() {
+            valid_credential_signatures_count += 1;
+        }
+    }
+
+    // Check if the total number of valid credential signatures meets the required
+    // threshold.
+    Ok(valid_credential_signatures_count >= credential_signatures_threshold.into())
+}
+
+/// Wrapper around the [`verify_account_signature`] function to simplify the
+/// verification for single-signer accounts.
+pub async fn verify_single_account_signature(
+    client: v2::Client,
+    signer: AccountAddress,
+    signature: Signature,
+    message: &Message<'_>,
+    bi: BlockIdentifier,
+) -> Result<bool, SignatureError> {
+    verify_account_signature(
+        client,
+        signer,
+        &AccountSignatures::singleton(signature),
+        message,
+        bi,
+    )
+    .await
+}
+
+/// Verifies account signatures similarly to the [`verify_account_signature`]
+/// function but does not perform any checks against the on-chain data such as:
+/// - if the signer account exists on chain.
+/// - if the credential or key indices from the `signature_data` map exist on
+///   chain.
+/// - if the thresholds on chain are followed.
+/// - if the public keys in the `signature_data` map match the public keys on
+///   chain.
+pub fn verify_account_signature_unchecked(
+    signer: AccountAddress,
+    signature_data: AccountSignaturesVerificationData,
+    message: &Message<'_>,
+) -> Result<bool, SignatureError> {
+    let message_hash = calculate_message_hash(message, signer);
+
+    for (_, credential) in signature_data.data {
```

Comment:
```suggestion
    for credential in signature_data.data.values() {
```

Same in the inner loop.